pkg/agent/billing.go

```diff
@@ -0,0 +1,181 @@
+package agent
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/google/uuid"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	vmapi "github.com/neondatabase/neonvm/apis/neonvm/v1"
+
+	"github.com/neondatabase/autoscaling/pkg/billing"
+	"github.com/neondatabase/autoscaling/pkg/util"
+)
+
+type BillingConfig struct {
+	URL                 string `json:"url"`
+	CollectEverySeconds uint   `json:"collectEverySeconds"`
+	PushEverySeconds    uint   `json:"pushEverySeconds"`
+	PushTimeoutSeconds  uint   `json:"pushTimeoutSeconds"`
+}
+
+type billingMetricsState struct {
+	historical map[billingMetricsKey]vmMetricsHistory
+	present    map[billingMetricsKey]vmMetrics
+	lastTime   *time.Time
+}
+
+type billingMetricsKey struct {
+	uid        types.UID
+	endpointID string
+}
+
+type vmMetricsHistory struct {
+	slices []metricsTimeSlice
+}
+
+type metricsTimeSlice struct {
+	metrics   vmMetrics
+	startTime time.Time
+	endTime   time.Time
+}
+
+type vmMetrics struct {
+	cpu uint16
+}
+
+const (
+	MetricNameCPU string = "cpu"
+	EndpointLabel string = "neon/endpoint-id"
+)
+
+func RunBillingMetricsColllector(backgroundCtx context.Context, conf *BillingConfig, store *util.WatchStore[vmapi.VirtualMachine]) error {
+	client := billing.NewClient(conf.URL, http.DefaultClient)
+
+	collectTicker := time.NewTicker(time.Second * time.Duration(conf.CollectEverySeconds))
+	defer collectTicker.Stop()
+	pushTicker := time.NewTicker(time.Second * time.Duration(conf.PushEverySeconds))
+	defer pushTicker.Stop()
+
+	state := billingMetricsState{
+		historical: make(map[billingMetricsKey]vmMetricsHistory),
+		present:    make(map[billingMetricsKey]vmMetrics),
+		lastTime:   nil,
+	}
+
+	state.collect(conf, store)
+
+	for {
+		select {
+		case <-collectTicker.C:
+			state.collect(conf, store)
+		case <-pushTicker.C:
+			batch := state.drainIntoBatch(&client)
+			if err := pushBillingEvents(conf, batch); err != nil {
+				return fmt.Errorf("Error pushing billing events: %w", err)
+			}
+		case <-backgroundCtx.Done():
+			// If we're being shut down, push the latests events we have before returning.
+			batch := state.drainIntoBatch(&client)
+			if err := pushBillingEvents(conf, batch); err != nil {
+				return fmt.Errorf("Error pushing billing events: %w", err)
+			}
+			return nil
+		}
+	}
+}
+
+func (s *billingMetricsState) collect(conf *BillingConfig, store *util.WatchStore[vmapi.VirtualMachine]) {
+	now := time.Now()
+
+	old := s.present
+	s.present = make(map[billingMetricsKey]vmMetrics)
+	for _, vm := range store.Items() {
+		endpointID, ok := vm.Labels[EndpointLabel]
+		if !ok {
+			continue
```

Comment:
is this "something went away" or "we have a pod that we don't know about"

Reply:
"we have a pod that we don't know about" — added a comment.